File: app/src/main/java/com/faendir/lightning_launcher/multitool/proxy/Image.java
package com.faendir.lightning_launcher.multitool.proxy;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.StringDef;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

/**
 * @author lukas
 * @since 04.07.18
 */
public interface Image extends Proxy {
    String TYPE_BITMAP = "BITMAP";
    int getWidth();

    int getHeight();

    @Type
    String getType();

    interface Class extends Proxy {
        static Class get(@NonNull Context context) {
            try {
                return ProxyFactory.lightningProxy(context.getClassLoader().loadClass("net.pierrox.lightning_launcher.script.api.Image"), Class.class);
            } catch (ClassNotFoundException e) {
                throw new RuntimeException(e);
            }
        }

        ImageBitmap createImage(int width, int height);

        Image createImage(String pkg, String name);
    }

    @Retention(RetentionPolicy.SOURCE)
    @StringDef({TYPE_BITMAP})
    @interface Type {
    }
}

File: app/src/main/java/com/faendir/lightning_launcher/multitool/proxy/Image.kt
package com.faendir.lightning_launcher.multitool.proxy

import android.content.Context
import androidx.annotation.StringDef

/**
 * @author lukas
 * @since 04.07.18
 */
interface Image : Proxy {
    val width: Int

    val height: Int

    @get:Type
    val type: String

    interface Class : Proxy {

        fun createImage(width: Int, height: Int): ImageBitmap

        fun createImage(pkg: String, name: String): Image

        companion object {
            operator fun get(context: Context): Class {
                try {
                    return ProxyFactory.lightningProxy(context.classLoader.loadClass("net.pierrox.lightning_launcher.script.api.Image"), Class::class.java)
                } catch (e: ClassNotFoundException) {
                    throw RuntimeException(e)
                }

            }
        }
    }

    @Retention(AnnotationRetention.SOURCE)
    @StringDef(TYPE_BITMAP)
    annotation class Type

    companion object {
        const val TYPE_BITMAP = "BITMAP"
    }
}
